metrics_sdk/lib/opentelemetry/sdk/metrics/meter_provider.rb

```diff
@@ -0,0 +1,134 @@
+# frozen_string_literal: true
+
+# Copyright The OpenTelemetry Authors
+#
+# SPDX-License-Identifier: Apache-2.0
+
+module OpenTelemetry
+  module SDK
+    # The Metrics module contains the OpenTelemetry metrics reference
+    # implementation.
+    module Metrics
+      # {MeterProvider} is the SDK implementation of {OpenTelemetry::Metrics::MeterProvider}.
+      class MeterProvider < OpenTelemetry::Metrics::MeterProvider
+        Key = Struct.new(:name, :version)
+        private_constant(:Key)
+
+        attr_reader :resource, :metric_readers
+
+        def initialize(resource: OpenTelemetry::SDK::Resources::Resource.create)
+          @mutex = Mutex.new
+          @meter_registry = {}
+          @stopped = false
+          @metric_readers = []
+          @resource = resource
+        end
+
+        # Returns a {Meter} instance.
+        #
+        # @param [String] name Instrumentation package name
+        # @param [optional String] version Instrumentation package version
+        #
+        # @return [Meter]
+        def meter(name, version = nil)
+          version ||= ''
+          if @stopped
+            OpenTelemetry.logger.warn 'calling MeterProvider#meter after shutdown, a noop meter will be returned.'
+            OpenTelemetry::Metrics::Meter.new
+          else
+            @mutex.synchronize { @meter_registry[Key.new(name, version)] ||= Meter.new(name, version, self) }
+          end
+        end
+
+        # Attempts to stop all the activity for this {MeterProvider}.
+        #
+        # Calls MetricReader#shutdown for all registered MetricReaders.
+        #
+        # After this is called all the newly created {Meter}s will be no-op.
+        #
+        # @param [optional Numeric] timeout An optional timeout in seconds.
+        # @return [Integer] Export::SUCCESS if no error occurred, Export::FAILURE if
+        #   a non-specific failure occurred, Export::TIMEOUT if a timeout occurred.
+        def shutdown(timeout: nil)
+          @mutex.synchronize do
+            if @stopped
+              OpenTelemetry.logger.warn('calling MetricProvider#shutdown multiple times.')
+              Export::FAILURE
+            else
+              start_time = OpenTelemetry::Common::Utilities.timeout_timestamp
+              results = @metric_readers.map do |metric_reader|
+                remaining_timeout = OpenTelemetry::Common::Utilities.maybe_timeout(timeout, start_time)
+                if remaining_timeout&.zero?
+                  Export::TIMEOUT
+                else
+                  metric_reader.shutdown(timeout: remaining_timeout)
+                end
+              end
+
+              @stopped = true
+              results.max || Export::SUCCESS
```

Comment:
When would we need this fallback case? (And if we didn't have any results, would we want `FAILURE` instead?)

Reply:
See below - this is only hit if we have no metric readers.